ext/opcache/shared_alloc_mmap.c

```diff
@@ -40,6 +40,340 @@
+            segments[j-1].start = temp.start;
+            segments[j-1].end = temp.end;
+        }
+    }
+}
+
+/* Combine adjacent/overlapped segments into one bigger segment,
+   and the result is placed in res array and its count is returned.
+*/
+static size_t combine_segments(struct map_seg_addresses* segments,
+                               size_t seg_count,
+                               struct map_seg_addresses* res,
+                               size_t max_res_count) {
+    size_t res_count = 0;
+
+    for (size_t i = 0; i < seg_count;) {
+        res[res_count].start = segments[i].start;
+
+        /* temp used to remember the biggest end address of adjacent segments. */
+        uintptr_t temp = segments[i].end;
+        size_t j = i + 1; /* next segment to check */
+
+        /* As long as segments[j] is adjacent or overlapped, we update temp value */
+        while (j < seg_count && segments[j].start <= temp) {
+            temp = max(temp, segments[j].end);
+            j++; /* check next segment */
+        }
+
+        if (res_count < max_res_count) {
+            res[res_count].end = temp;
+            res_count++;
+        } else { /* error: index overflow */
+            return 0;
+        }
+
+        i = j; /* jump to the next unchecked segment */
+    }
+
+    return res_count;
+}
+
+/* Search all candidate buffer addresses in order for opcache/jit buffer.
+   All available addresses are stored in candidates array; candidate count
+   is returned.
+*/
+static size_t search_candidates(struct map_seg_addresses* combined_segments,
+                    size_t seg_count,
+                    size_t requested_size,
+                    struct map_seg_addresses* php_text_segment,
+                    uintptr_t* candidates,
+                    size_t max_candidates_count)
+{
+    uintptr_t prev_seg_end = 0;   /* previous segment end address */
+    uintptr_t candidate = 0;      /* candidate address */
+    size_t candidate_count = 0;
+
+#if defined(MAP_HUGETLB)
+    size_t reserved_align_space = huge_page_size;
+#else
+    size_t reserved_align_space = ordinary_page_size;
+#endif /* MAP_HUGETLB */
+
+    /* No need to move buffer closer to PHP .text segment, if
+       1) requested buffer size is equal to or greater than 4GB,
+       2) PHP .text segment is under 4GB memory address
+    */
+    if (requested_size >= size_4GB || php_text_segment->start < addr_4GB)
+    {
+        return 0;
+    }
+
+    /*  PHP segments diagram
+
+               [seg0]     [seg1]      [php .text][heap]      [seg3]
+        [hole0]     [hole1]     [hole2]                [hole3]
+
+        We only search any candidates BEFORE PHP .text segment. E.g., [hole0],
+        [hole1], and [hole2] might be good candidates. Here is why:
+        in standalone PHP, we find that [heap] is closely following PHP .text
+        segment and our buffer might block heap growth if we use [hole3].
+    */
```

Comment:
When PHP is a shared library this may not be true (the .text segment may be after the heap). This may not be a problem though.

Reply:
Will add some more comments to make it clearer.